### PR TITLE
Libceed scipy paper

### PR DESCRIPTION
**Pull request by valeriabarra**
Automated PR comment:
------------------------------------------
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

My comment:
------------------------------------------
To compile:
```
$ cd scipy_proceedings
```

```
$ ./make_paper.sh papers/libceed-paper/
```

Output in PDF can be found in `scipy_proceedings/output/libceed-paper/paper.pdf`

### CODE REVIEW

**Comment by valeriabarra** [comment]
Update:
I have updated the file according to the conference proceeding template.

I have struggled for quite some time to make figure labels work according to their [example](https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst) but it did not work for me, I don't know why. I removed the labels for now, but of course this way we have undefined references when referring to the figures in the text.

File: papers/libceed-paper/paper.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the CeedVector structure are associated to arrays defined via the numpy package. In fact, since libCEED heavily relies on pointers and arrays to handle the data, a Python structure that resembles the C arrays is needed. In details, numpy arrays allow this correspondence obtained by passing the numpy array memory address as pointers to the libCEED C API.
```
Comment (jeremylt) [inline papers/libceed-paper/paper.rst]: I'm not sure how much effort it would be, but we could try using numpy or gnumpy so we have clearer CUDA + Python support? If its easy to add, it would be worth mentioning. I can look into it.

File: papers/libceed-paper/paper.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+:bibliography: bibliography
+
+----------------------------------------------------------------------------------
+High-performance operator evaluations with ease of use: libCEED's Python interface
+----------------------------------------------------------------------------------
+
+.. class:: abstract
+
+   libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. libCEED offers implementations, selectable at runtime, tuned for a variety of computational device types, including CPUs and GPUs. libCEED’s purely algebraic framework can unobtrusively be integrated in new and legacy software to provide performance portable applications. In this work, we present libCEED's newly available Python interface that opens up new strategies for parallelism and scaling in high-performance Python, without having to compromise ease of use.
```
Comment (YohannDudouit) [inline papers/libceed-paper/paper.rst]: I feel like we should mention that libCEED is a "Matrix-Free Finite Element library".

File: papers/libceed-paper/paper.rst [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+----------------------------------------------------------------------------------
+
+LibCEED's API provides the local action of the linear or nonlinear operator without assembling its sparse representation. Let us define the global operator as
+
+.. math::
+   :label: eq-operator-decomposition
+
+   A = P^T \underbrace{G^T B^T D B G}_{\text{libCEED}} P \, ,
+
+where :math:`P` is the parallel process decomposition operator (external to ``libCEED``) in which the degrees of freedom (DOFs) are scattered to and gathered from the different compute devices. The operator denoted by :math:`A_L = G^T B^T D B G` gives the local action on a compute node or process, where :math:`G` is a local element restriction operation that localizes DOFs based on the elements, :math:`B` defines the action of the basis functions (or their gradients) on the nodes, and :math:`D` is the user-defined pointwise function describing the physics of the problem at the quadrature points, also called the QFunction (see Fig. :ref:`fig-operator-decomp`).
```
Comment (jeremylt) [inline papers/libceed-paper/paper.rst]: Perhaps here it would be appropriate to mention "This parallel decomposition is out of scope for the libCEED API and needs to be managed by the user or an additional package, such as petsc4py."

File: papers/libceed-paper/paper.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+.. class:: keywords
+
+   High-performance Python, performance portability, scalability, parallelism
+
+Introduction
+----------------------------------------------------------------------------------
+
+libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. Historically, conventional high-order finite element methods were rarely used for industrial problems because the Jacobian rapidly loses sparsity as the order is increased, leading to unaffordable solve times and memory requirements :cite:`brown2010`. This effect typically limited the order of accuracy to at most quadratic, especially because they are computationally advantageous in terms of floating point operations (FLOPS) per degree of freedom (DOF)---see :ref:`fig-assembledVsmatrixfree` ---, despite the fast convergence and favorable
```
Comment (YohannDudouit) [inline papers/libceed-paper/paper.rst]: I find the argument about Jacobians a bit too restrictive, I think that in general high-order impacts the sparsity (if we don't assume any specific basis functions)

File: papers/libceed-paper/paper.rst (hunk 1 above)
Comment (jedbrown) [reply papers/libceed-paper/paper.rst]: I think JAX is probably the most interesting approach for device arrays, and it would bring with it the best in Python automatic differentiation.

File: papers/libceed-paper/paper.rst (hunk 3 above)
Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: We say (external to ``libCEED``) but sure

File: papers/libceed-paper/paper.rst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+stability properties offered by higher order discretizations. Nowadays, high-order numerical methods are widely used in Partial Differential Equation (PDE) solvers, but software packages that provide high-performance implementations have often been special-purpose and intrusive. In contrast, libCEED is light-weight, minimally intrusive, and very versatile. When high-order finite/spectral element discretizations are used, the resulting sparse matrix representation of a global operator is computationally expensive, with respect to both the memory transfer and floating point operations needed for its evaluation. libCEED's Application Programming Interface (API) provides the local action of an operator (linear or nonlinear) without assembling its sparse representation. The purely algebraic nature of libCEED allows efficient operator evaluations (selectable at runtime) and offers matrix-free preconditioning ingredients. While libCEED’s focus is on high-order finite elements, the approach with which it is designed is algebraic and thus applicable to other discretizations in factored form.
```
Comment (YohannDudouit) [inline papers/libceed-paper/paper.rst]: Maybe we should cite Orszag's paper: "Spectral Methods for Problems in Complex Geometries" as a justification that sparse matrices are not a good representation for high-order.

Comment (YohannDudouit) [reply papers/libceed-paper/paper.rst]: We might also want to mention that algebraic decomposition is more general than a matrix representation of the finite element operator, something like:
This decomposition also presents the benefit that it can represent non-linear finite element operators as well.

File: papers/libceed-paper/paper.rst [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+.. figure:: libCEED.png
+   :align: center
+   :figclass: bht
+
+   Operator decomposition.
+
+The mathematical formulation of QFunctions, described in weak form, is fully separated from the parallelization and meshing concerns. In fact, QFunctions, which can either be defined by the user or selected from a gallery of available built-in functions in the library, are pointwise functions that do not depend on element resolution, topology, or basis degree (selectable at run time). This easily allows :math:`hp`-refinement studies (where :math:`h` commonly denotes the average element size and :math:`p` the polynomial degree of the basis functions in 1D) and :math:`p`-multigrid solvers. libCEED also supports composition of different operators for multiphysics problems and mixed-element meshes (see Fig. :ref:`fig-schematic`). Currently, user-defined QFunctions are written in C and must be precompiled before running the Python script. The ultimate goal is for users to write only Python code. This will be achieved in the near future by using the Numba high-performance Python compiler.
```
Comment (jeremylt) [inline papers/libceed-paper/paper.rst]: Perhaps
"Currently, user-defined QFunctions are written in C and must be precompiled as a foreign function library and loaded via ctypes."
and (if we show easy GPU support, which I think would be cool)
"However, these C QFunctions allow users to compute on GPU devices that libCEED supports."
(If OCCA is merged by then, that will be Nvidia + AMD)

Aside - Numba won't offer single source though, right? You can run Numba + CUDA, but it won't work with our CUDA backends.

File: papers/libceed-paper/paper.rst [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the CeedVector structure are associated to arrays defined via the numpy package. In fact, since libCEED heavily relies on pointers and arrays to handle the data, a Python structure that resembles the C arrays is needed. In details, numpy arrays allow this correspondence obtained by passing the numpy array memory address as pointers to the libCEED C API.
+
+libCEED's API
+----------------------------------------------------------------------------------
+
+LibCEED's API provides the local action of the linear or nonlinear operator without assembling its sparse representation. Let us define the global operator as
```
Comment (YohannDudouit) [inline papers/libceed-paper/paper.rst]: I find the word "local" confusing here, I would reword like this:
"LibCEED's API provides the action of the linear or nonlinear Finite Element operator without assembling its sparse representation."

File: papers/libceed-paper/paper.rst (hunk 1 above)
Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: If we can run some tests to verify, then we already have libCEED + Python + CUDA and can say so here, as you can use jax.numpy in place of numpy, if I'm understanding correctly.

File: papers/libceed-paper/paper.rst (hunk 3 above)
Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: Right, I mostly was thinking that we could highlight how libCEED can plug in with other packages that are already popular.

File: papers/libceed-paper/paper.rst (hunk 1 above)
Comment (jedbrown) [reply papers/libceed-paper/paper.rst]: FFI aspects of `jax.numpy` are surely different (at least for device arrays).

BTW, JAX is sort of a combined successor to numba and autograd (in the sense that developers moved from both of those projects to JAX); it does both JIT and AD, both with device support. There might be very little work needed to run on `jax.numpy` arrays, but someone would need to experiment.

File: papers/libceed-paper/paper.rst (hunk 7 above)
Comment (YohannDudouit) [reply papers/libceed-paper/paper.rst]: I think it would be a bit better to give the intuition that instead of forming a single operator using a sparse matrix, we're going to use a product of operations to decompose our operator that will be applied __sequentially__.

File: papers/libceed-paper/paper.rst [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+High-performance operator evaluations with ease of use: libCEED's Python interface
+----------------------------------------------------------------------------------
+
+.. class:: abstract
+
+   libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. libCEED offers implementations, selectable at runtime, tuned for a variety of computational device types, including CPUs and GPUs. libCEED’s purely algebraic framework can unobtrusively be integrated in new and legacy software to provide performance portable applications. In this work, we present libCEED's newly available Python interface that opens up new strategies for parallelism and scaling in high-performance Python, without having to compromise ease of use.
+
+.. class:: keywords
+
+   High-performance Python, performance portability, scalability, parallelism
```
Comment (YohannDudouit) [inline papers/libceed-paper/paper.rst]: Don't we want "Finite Element" and "High order" in the keywords?

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Thanks. I retrieved the ones used on easychair for the abstract submission. I don't recall now if there was a list to choose from. I think we could add some more, and hopefully the reviewers won't complain if they do not exactly match the abstract submission ones.

File: papers/libceed-paper/paper.rst (hunk 4 above)
Comment (YohannDudouit) [reply papers/libceed-paper/paper.rst]: We could also cite Orszag's paper with Jed's paper.

File: papers/libceed-paper/paper.rst (hunk 7 above)
Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Attempted in [8187f04](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/8187f0446ea4ca87e782424c4e7d6d0514e9a246) please let me know if it sounds good

File: papers/libceed-paper/paper.rst (hunk 4 above)
Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: I think it is neater this way. One citation specifically refers to the sparsity and one to the cost.

File: papers/libceed-paper/paper.rst (hunk 6 above)
Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Attempted in [471a5cc](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/471a5cc68c77240ad12631faa9ccb5c17c6e93cd) (altough the diff is hard to read, even though I used a single line per paragraph, as discussed in the documentation PR)

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: I'm not sure about your aside to be honest. Need to learn more

Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: Right now, the CUDA/MAGMA backends take the single source QFunctions in C source files and compile them as device kernels, with appropriate change to the CEED_QFUNCTION macro. If I understand the capabilities correctly, we can use Numba to make a function pointer for the CPU backends but that code won't be compatible with single source QFunctions. We might be able to make a backend that works with numba CUDA, though I think it would take more work than on the CPU side as having the QFunction source is more integral to the CUDA backends.

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: I see...

Comment (jedbrown) [reply papers/libceed-paper/paper.rst]: We should refer to JAX instead of (or in addition to) Numba. Notes here say sharing of device arrays is supported. https://jax.readthedocs.io/en/latest/CHANGELOG.html?highlight=__cuda_array_interface__#jax-0-1-59-february-11-2020

I believe the cuda-shared and magma backends would be able to use JAX or Numba qfunctions, but cuda-gen would not because the *.cu source is not available (so it would at least need significant JIT shenanigans, if supportable at all).

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Mentioned both possibilities in [c5a9041](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/c5a9041c8f26486f687aaad0b68c75bde90ed197)

File: papers/libceed-paper/paper.rst (hunk 7 above)
Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: We perhaps don't want to say the word sequentially? The exact implementation is backend specific. In some cases we apply suboperators partially, like in the CPU backends. What about:

Instead of forming a single operator using a sparse matrix representation, libCEED composes the subcomponents described in equation (:ref:`eq-operator-decomposition`) to provide the action of the operator in a fashion tuned for the hardware available to the user.

Or something like that?

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Attempted again in [204b47d](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/204b47d14efbdf3e04cfdcd0f957a72c2c855eec)

File: papers/libceed-paper/paper.rst (hunk 6 above)
Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: /gpu/cuda/ref /gpu/cuda/reg and /gpu/cuda/shared would still require some modification to replace the compilation of the QF with getting the kernel from JAX/Numba but it looks easier. I agree that I'm not sure if /cuda/gen is possible.

File: papers/libceed-paper/paper.rst [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
...The mathematical formulation of QFunctions, described in weak form, is fully separated from the parallelization and meshing concerns. In fact, QFunctions, which can either be defined by the user or selected from a gallery of available built-in functions in the library, are pointwise functions that do not depend on element resolution, topology, or basis degree (selectable at run time). This easily allows :math:`hp`-refinement studies (where :math:`h` commonly denotes the average element size and :math:`p` the polynomial degree of the basis functions in 1D) and :math:`p`-multigrid solvers. libCEED also supports composition of different operators for multiphysics problems and mixed-element meshes (see Fig. :ref:`fig-schematic`). Currently, user-defined QFunctions are written in C and must be precompiled as a foreign function library and loaded via _ctypes_. The single-source C QFunctions allow users to equally compute on CPU or GPU devices, supported by libCEED. The ultimate goal is for users to write only Python code. This will be achieved in the near future by using the Numba high-performance Python compiler or Google's extensible system for composable function transformations, Jax.
```
Comment (jeremylt) [inline papers/libceed-paper/paper.rst]: I think JAX is all caps

File: papers/libceed-paper/paper.rst (hunk 7 above)
Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: I like it

File: papers/libceed-paper/paper.rst (hunk 9 above)
Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: True. Thanks

File: papers/libceed-paper/paper.rst (hunk 7 above)
Comment (YohannDudouit) [reply papers/libceed-paper/paper.rst]: All the implementations are sequential, but are more or less local.

Comment (jeremylt) [reply papers/libceed-paper/paper.rst]: Sort of?
In the blocked CPU backends, some restrictions are fully applied and some are partially applied.
Then in a loop over element batches, active restrictions are applied to a portion of the inputs, the basis operations are applied to a portion of the elements, qfunction is applied for a portion of the elements, the output basis operations are applied to a portion of the elements, and output restrictions are applied to a portion of the elements.

I personally wouldn't call that a sequential application of objects in the decomposition in the way the that all restrictions, all basis operations, a qfunction call, all transpose operations, and all transpose restrictions are applied fully in sequence in non-optimized backends.

Comment (YohannDudouit) [reply papers/libceed-paper/paper.rst]: I wrote "more or less local" the operations are still applied sequentially...

**Comment by valeriabarra** [comment]
I've been trying to have `:label:` work in figure captions as they have in their [example](https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst) but I don't know if it's a problem of the `python3-docutils` package for my distribution or something else. But whenever I try to add a figure `:label:` and compile (for instance `` :label:`fig-assembledVsmatrixfree` ``, I get the following error:

```
! Argument of \@gobble has an extra }.
<inserted text> 
                \par 
l.170 ...DUrole{label}{fig-assembledVsmatrixfree}}

```
Please let me know if any of you experience a similar issue.

**Comment by valeriabarra** [comment]
If, instead I use the simple referencing in `rst`, that is, the `.. _fig-label:` before the `.. figure::` directive, the compilation completes well and creates a PDF, but the figure has the missing reference anyways.

**Comment by valeriabarra** [comment]
Hooray! I was finally able to solve my `docutils` distribution issue and now I am able to add and see all figure references. The only thing left that I see not rendered properly now is the `---` for em dash

**Comment by valeriabarra** [comment]
Hey @jedbrown @YohannDudouit @jeremylt  I have been adding some better descriptions and code snippets of the Python interface. 

Can anyone give me any input on how to proceed? The deadline is in 2 days. We need to merge this here before being able to create the PR on the conference website for the submission on Friday. Many thanks to all.

File: papers/libceed-paper/paper.rst [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+.. figure:: libCEEDBackends.png
+
+   The role of libCEED as a lightweight, portable library which provides a low-level API for efficient, specialized implementations. libCEED allows different applications to share highly optimized discretization kernels. :label:`fig-libCEED-backends`
+
```
Comment (jeremylt) [inline papers/libceed-paper/paper.rst:58]: I think we can drop " and C interface" part of the first sentence since we don't really discuss the C API past mentioning that the Python interface is a wrapper for the C library. That would help address the reviewers that were concerned that this talks too much about C.

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst:58]: Reworded. I meant we'd introduce the API.

File: papers/libceed-paper/paper.rst [hunk 11]
```diff
@@ -34,16 +34,16 @@ High-performance operator evaluations with ease of use: libCEED's Python interfa
...
+Conventional high-order finite element methods have seen very limited use for industrial problems because the matrix describing the action of the operator loses sparsity as the order is increased :cite:`Orszag1980`, leading to unaffordable solve times and memory requirements :cite:`brown2010`. This effect typically limited the polynomial basis degree to at most quadratic, especially because they are computationally advantageous [TODO: did you mean to say it gave up a lot?] in terms of floating point operations (FLOPs) per degree of freedom (DOF)—see Fig. :ref:`fig-assembledVsmatrix-free`—despite the fast convergence and favorable stability properties offered by higher order discretizations. Nowadays, high-order numerical methods, such as the spectral element method (SEM)—a special case of nodal p-Finite Element Method (FEM) that can reuse the interpolation nodes for quadrature—are employed [TODO: can say they are popular is some domains; cite Nek5000 and MFEM], especially with (nearly) affine elements, because linear constant coefficient problems can be very efficiently solved [TODO: needs to be *separable*; cite sum factorization].
```
Comment (valeriabarra) [inline papers/libceed-paper/paper.rst]: Thanks @jedbrown I don't understand one of your questions. What did you mean when you said `[TODO: did you mean to say it gave up a lot?] ` Did you refer to the fact that quadratic elements are advantageous in terms of FLOPs per DOFs as explained in the sentence?

Comment (jedbrown) [reply papers/libceed-paper/paper.rst]: It limits the degree **because** "they are computationally advantageous"? Something is missing/unclear in this sentence.

Comment (valeriabarra) [reply papers/libceed-paper/paper.rst]: Thanks. I think I addressed it in [d57be80](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/d57be804c51c6ff69e303455465d90f31c8fc630)

**Comment by valeriabarra** [comment]
Thanks you all for your inputs and suggestions. I made some final edits. 

Please let me know if we can merge this and then I can create the PR in the conference's repo.
